Add OpenStack Security Group Rule Module

emonty opened the pull request.

emonty committed:
Add OpenStack Security Group Rule module

emonty committed:
Fix example code syntax

dguerri committed:
Fix invalid syntax in openstack_module_kwargs call

dagnello committed:
Minor fixes for os_security_group_rule module

Was not able to use this module as it was.  The changes submitted
resolved the issues I ran into in order to get it working.

dagnello committed:
Minor fixes for os_security_group_rule module

Was not able to use this module as it was.  The changes submitted
resolved the issues I ran into in order to get it working.

Commit:
Update rules mode for latest shade

Shade 0.7.0 normalized the security group data that is returned,
when using nova, to look more like neutron security group data.
This adjusts for that change.

Commit:
Fix id value reference

Commit:
Recongnize None and -1 port equivalency

shade 0.7.0 represents disabled min/max ports as None (in the
neutron style) rather than -1. Recognize this as the same as -1.

Commit:
Update secgroup rules module for latest shade

This allows the rules module to work against either nova or neutron
for handling security groups. New parameters for 'direction' and
'ethertype' are added.

Check mode is supported with this version.

emonty committed:
Return rule object

Commit:
Fix syntax error

Commit:
Compare ports as strings

Ports as returned from shade are ints. They are strings as they come
in to the module.

dagnello committed:
Resolving issues in rule comparison algorithm

Port range min/max values are at times represented as string and
compared to int equivalents.  This fix explicitly ensures all
port range values are ints for proper comparisons.

emonty committed:
Use int in the parameter list instead of casting

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,214 @@
+    argument_spec = openstack_full_argument_spec(
+        security_group   = dict(required=True),
+        protocol         = dict(default='tcp',
+                                choices=['tcp', 'udp', 'icmp']),
```
At least when dealing with a Neutron end point, a protocol is not required. The default, given that this is optional according to the API spec, should be None, or Null. It's how one can create a rule that spans all IP protocols. This module should allow for protocol to be not specified and thus default to None.

Shrews replied:
Fixed! Thanks for pointing this out. Shade has been changed to toss an exception if Nova is being used and no protocol is supplied.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,214 @@
+        protocol         = dict(default='tcp',
+                                choices=['tcp', 'udp', 'icmp']),
+        port_range_min   = dict(required=True, type='int'),
+        port_range_max   = dict(required=True, type='int'),
```
With Neutron port ranges are optional, with a default of Null, which would span all ports. As such this module should not require port_range_min/max and default to None.

Shrews replied:
Fixed.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,214 @@
+        # TODO(mordred): Make remote_group handle name and id
+        remote_group     = dict(required=False, default=None),
+        ethertype        = dict(default='IPv4',
+                                choices=['IPv4', 'IPv6']),
```
Technically the API doesn't require IPv4 or IPv6 as a ethertype, the whole argument is listed as Optional, but behavior is unknown as to what it'll do. Food for thought.

emonty replied:
Let's none of us spend too much time thinking about why it would be optional or we might do harm to someone.

Commit:
Change required parameters for rules module

The ports and protocol are no longer required (and now depends on
a new version of shade).

Shrews commented:
Let's keep the needs_revision label on this until a newer version of shade is released.

omgjlk commented on cloud/openstack/os_security_group_rule.py:
```diff
@@ -0,0 +1,257 @@
+    for rule in secgroup['security_group_rules']:
+        # No port, or -1, will be returned from shade as None
+        if rule['port_range_min'] is None:
+            rule_port_range_min = -1
```
This is breaking comparison on rules that do not have port ranges. Rules without port ranges are coming back from shade as None and this code is setting it to -1, however rules being passed to the module are also coming through as None and not being converted here. Can we do away with the conversion to -1, or match on both -1 AND None?

Shrews replied:
I believe you need a shade patch I created for this. Will verify tomorrow. 

Sent from my iPhone

> On Jun 30, 2015, at 6:13 PM, Jesse Keating notifications@github.com wrote:
> 
> In cloud/openstack/os_security_group_rule.py:
> 
> > +def _find_matching_rule(module, secgroup):
> > -    """
> > -    Find a rule in the group that matches the module parameters.
> > -    :returns: The matching rule dict, or None if no matches.
> > -    """
> > -    protocol = module.params['protocol']
> > -    port_range_min = module.params['port_range_min']
> > -    port_range_max = module.params['port_range_max']
> > -    remote_ip_prefix = module.params['remote_ip_prefix']
> > -    ethertype = module.params['ethertype']
> > -    direction = module.params['direction']
> >   +
> > -    for rule in secgroup['security_group_rules']:
> > -        # No port, or -1, will be returned from shade as None
> > -        if rule['port_range_min'] is None:
> > -            rule_port_range_min = -1
> >   This is breaking comparison on rules that do not have port ranges. Rules without port ranges are coming back from shade as None and this code is setting it to -1, however rules being passed to the module are also coming through as None and not being converted here. Can we do away with the conversion to -1, or match on both -1 AND None?
> 
> —
> Reply to this email directly or view it on GitHub.

Shrews replied:
@j2sol Yep, you were correct. It was easier to just convert incoming -1 values to None so that they match what is returned from shade. I added a lot of inline documentation to describe what the various actors are doing and how we need to do comparisons. I also created a playbook to test _several_ various scenarios, and tested against both Nova and Neutron, and it looks pretty solid now. Hopefully another set of eyes will help catch anything I may have missed.

omgjlk commented:
Other than the -1/None issue this module is working as expected for our environment.

gregdek commented:
Looks like this one is bitten by our keypair docs failure from a few days back -- @emonty can you rebase?

Commit:
Correct port matching logic

Port matching logic did not take into account recent shade change
to equate (None, None) to (1, 65535) when Nova is the backend.

Also, this encapsulates the port matching logic into a single function
and heavily documents the logic.

omgjlk commented:
This works for me now, and passes tests. Are we still waiting on a shade version?

Also the doc string needs to be updated indicating that the port min/max is not required.

emonty committed:
Update docstring to show port ranges as optional

emonty commented:
@j2sol nope- shade is current and in good shape. @gregdek rebased and ready to fly

omgjlk commented:
:+1:
